src/plan_add_hashagg.c

```diff
@@ -0,0 +1,595 @@
+	if (IsA(left, Const))
+		nonconst = right;
+	else if (IsA(right, Const))
+		nonconst = left;
+	else
+		return INVALID_ESTIMATE;
+
+	/* adding or subtracting a constant doesn't affect the range */
+	if (function_name[0] == '-' || function_name[0] == '+')
+		return estimate_max_spread_expr(root, nonconst);
+
+	return INVALID_ESTIMATE;
+}
+
+/* estimate the max spread (max(value)-min(value)) of the expr */
+static double
+estimate_max_spread_expr(PlannerInfo *root, Expr *expr)
+{
+	if (IsA(expr, Var))
+		return estimate_max_spread_var(root, (Var *) expr);
+	if (IsA(expr, OpExpr))
+		return estimate_max_spread_opexpr(root, (OpExpr *) expr);
+	return INVALID_ESTIMATE;
+}
+
+/* Return an estimate for the number of groups formed when expr is divided
+   into intervals of size interval_priod. */
+static double
+custom_group_estimate_expr_interval(PlannerInfo *root, Expr *expr, double interval_period)
+{
+	double		max_period;
+
+	if (interval_period <= 0)
+		return INVALID_ESTIMATE;
+
+	max_period = estimate_max_spread_expr(root, expr);
+	if (!IS_VALID_ESTIMATE(max_period))
+		return INVALID_ESTIMATE;
+
+	return clamp_row_est(max_period / interval_period);
+}
+
+/* For time_bucket this estimate currently works by seeing how many possible
+ * buckets there will be if the data spans the entire hypertable. Note that
+ * this is an overestimate.
+ * */
+static double
+custom_group_estimate_time_bucket(PlannerInfo *root, FuncExpr *expr, double path_rows)
+{
+	Node	   *first_arg = eval_const_expressions(root, linitial(expr->args));
+	Expr	   *second_arg = lsecond(expr->args);
+	Const	   *c;
+	Interval   *interval;
+
+	if (!IsA(first_arg, Const))
+		return INVALID_ESTIMATE;
+
+	c = (Const *) first_arg;
+	interval = DatumGetIntervalP(c->constvalue);
+	return custom_group_estimate_expr_interval(root, second_arg, (double) get_interval_period_approx(interval));
+}
+
+/* For date_trunc this estimate currently works by seeing how many possible
+ * buckets there will be if the data spans the entire hypertable. Note that
+ * this is an overestimate.
+ * */
+static double
+custom_group_estimate_date_trunc(PlannerInfo *root, FuncExpr *expr, double path_rows)
+{
+	Node	   *first_arg = eval_const_expressions(root, linitial(expr->args));
+	Expr	   *second_arg = lsecond(expr->args);
+	Const	   *c;
+	text	   *interval;
+
+	if (!IsA(first_arg, Const))
+		return INVALID_ESTIMATE;
+
+	c = (Const *) first_arg;
+	interval = DatumGetTextPP(c->constvalue);
+	return custom_group_estimate_expr_interval(root, second_arg, (double) date_trunc_interval_period_approx(interval));
+}
+
+/* if performing integer division number of groups is less than the spread divided by the divisor.
+ * Note that this is an overestimate. */
+static double
+custom_group_estimate_integer_division(PlannerInfo *root, Oid opno, Node *left, Node *right)
+{
+	char	   *function_name = get_opname(opno);
+
+	/* only handle division */
+	if (strlen(function_name) == 1 && function_name[0] == '/' && IsA(right, Const))
```

Comment:
I guess this could also be `function_name[0] == '/' && function_name[1] == '\0' && ...`. Just sayin'...